packages/rollup-core/src/app/data/consumers/canonical-chain-batch-submitter.ts

```diff
@@ -78,39 +79,88 @@ export class CanonicalChainBatchSubmitter extends ScheduledTask {
       const msg = `Received tx batch to send in ${
         batchSubmission.status
-      } instead of ${
-        BatchSubmissionStatus.QUEUED
+      } instead of ${BatchSubmissionStatus.QUEUED} or ${
+        BatchSubmissionStatus.SUBMITTING
       }. Batch Submission: ${JSON.stringify(batchSubmission)}.`
       log.error(msg)
       throw new UnexpectedBatchStatus(msg)
     }
 
-    try {
-      const batchBlockNumber = await this.getBatchSubmissionBlockNumber()
+    if (await this.shouldSubmitBatch(batchSubmission)) {
+      try {
+        const batchBlockNumber = await this.getBatchSubmissionBlockNumber()
 
-      await this.validateBatchSubmission(batchSubmission, batchBlockNumber)
+        await this.validateBatchSubmission(batchSubmission, batchBlockNumber)
 
-      const txHash: string = await this.buildAndSendRollupBatchTransaction(
-        batchSubmission,
-        batchBlockNumber
-      )
-      if (!txHash) {
+        const txHash: string = await this.buildAndSendRollupBatchTransaction(
+          batchSubmission,
+          batchBlockNumber
+        )
+        if (!txHash) {
+          return false
+        }
+        batchSubmission.submissionTxHash = txHash
+      } catch (e) {
+        logError(log, `Error validating or submitting rollup tx batch`, e)
+        if (throwOnError) {
+          // this is only used by testing
+          throw e
+        }
         return false
       }
-      return this.waitForProofThatTransactionSucceeded(txHash, batchSubmission)
+    }
+
+    try {
+      return this.waitForProofThatTransactionSucceeded(
+        batchSubmission.submissionTxHash,
+        batchSubmission
+      )
     } catch (e) {
-      logError(log, `Error validating or submitting rollup tx batch`, e)
-      if (throwOnError) {
-        // this is only used by testing
-        throw e
-      }
+      logError(
+        log,
+        `Error waiting for canonical tx chain batch ${batchSubmission.batchNumber} with hash ${batchSubmission.submissionTxHash} to succeed!`,
+        e
+      )
       return false
     }
   }
 
+  protected async shouldSubmitBatch(batchSubmission): Promise<boolean> {
+    return (
+      batchSubmission.status === BatchSubmissionStatus.QUEUED ||
+      !(await isTxSubmitted(
+        this.canonicalTransactionChain.provider,
+        batchSubmission.submissionTxHash
+      ))
+    )
+  }
+
+  protected async getSignedRollupBatchTx(
+    txsCalldata: string[],
+    timestamp: number,
+    batchBlockNumber: number,
+    startIndex: number
+  ): Promise<string> {
+    return getSignedTransaction(
+      this.canonicalTransactionChain,
+      'appendSequencerBatch',
+      [txsCalldata, timestamp, batchBlockNumber, startIndex],
+      this.submitterWallet
+    )
+  }
+
+  protected async getBatchSubmissionBlockNumber(): Promise<number> {
+    // TODO: This will eventually be part of the output metadata from L2 tx outputs
+    //  Need to update geth to have this functionality so this is a mock for now
```

Comment:
👍 ah yes we definitely need this. Got to double check that we have it represented as a ticket